app/src/cc/arduino/packages/formatter/AStyle.java

```diff
@@ -76,15 +80,64 @@ public void init(Editor editor) {
 
   @Override
   public void run() {
-    String originalText = editor.getCurrentTab().getText();
-    String formattedText = aStyleInterface.AStyleMain(originalText, formatterConfiguration);
 
-    if (formattedText.equals(originalText)) {
-      editor.statusNotice(tr("No changes necessary for Auto Format."));
-      return;
-    }
+    SketchTextArea textArea = editor.getCurrentTab().getTextArea();
+
+    String originalText = textArea.getSelectedText();
+
+    // If no selection use all file
+    if (originalText == null || originalText.isEmpty()) {
+
+      String formattedText = aStyleInterface.AStyleMain(textArea.getText(), formatterConfiguration);
+      editor.getCurrentTab().setText(formattedText);
+
+    } else {
+      try {
+
+        // apply indentation control keywords.
+        String FORMAT_ON = "\n// *INDENT-ON* DYN\n";
+        String FORMAT_OFF = "\n// *INDENT-OFF* DYN\n";
 
-    editor.getCurrentTab().setText(formattedText);
+        RSyntaxDocument content = (RSyntaxDocument) textArea.getDocument();
+
+        textArea.beginAtomicEdit();
+
+        int selStart = editor.getCurrentTab().getSelectionStart();
+        int selEnd = editor.getCurrentTab().getSelectionStop();
+        int lineStart = textArea.getLineOfOffset(selStart);
+        int lineEnd = textArea.getLineOfOffset(selEnd);
+
+        // Calculate offsets from begin and end of each line.
+        int fristLineOffset = textArea.getLineStartOffset(lineStart);
+        int lastLineOffset = textArea.getLineEndOffset(lineEnd);
+
+        // inserts change the length, use this to calculate new positios.
+        int offLength = FORMAT_OFF.length();
+        int onLength = FORMAT_ON.length();
+
+        content.insertString(0, FORMAT_OFF, null);
+        content.insertString(fristLineOffset + offLength, FORMAT_ON, null);
+        content.insertString(lastLineOffset + offLength + onLength, FORMAT_OFF,null);
+        originalText = content.getText(0, content.getLength());
+
+        String formattedText = aStyleInterface.AStyleMain(originalText, formatterConfiguration);
+
+        // Remove format tags
+        formattedText = formattedText.replaceAll(Pattern.quote(FORMAT_OFF), "");
+        formattedText = formattedText.replaceAll(Pattern.quote(FORMAT_ON), "");
```

Comment:
This seems too general: If the code already contained these format off tags, they would now be removed. I would suggest to remember the offset where the tag was inserted (offset from the end for the third tag), and remove it only from there. As a sanity check, you could check that astyle has indeed left all text up to and including the second tag (and from the third tag to the end) indeed unmodified, otherwise your offsets would not be valid anymore.

Reply:
No problem, i put keyword 'DYN' to allow this and not conflict with pre-existing format tags

Reply:
Ah, I see. I would still think this is somewhat fragile (a user might also put "DYN" in there), but I guess it works well enough in practice. On advantage of doing some offset calculations is that you can at the same time verify that nothing outside of the selection was changed.